feat(input): add support for input[type=range]

Narretz opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
feature

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

the input[type=range] behavior is the same of an input[type=number]
with min=0, max=100 and step=1 as defaults

Closes #5892
Closes #9715

Narretz commented:
The biggest issue is that browser that implement range set the element value to a valid value if you try to set an invalid value. Two cases: you try to set a non-number: input value is set to 50. You try to set a value that exceeds max or is lower than min: value is set to max / min value.

This is problematic for two reasons:
- When we try to set the input value from the scope, as it creates a mismatch between input and scope value.
- null / undefined models are set to 50, which means the input can never have the required error set, even though from an application perspective the model is still null / undefined

In the PR, I have changed it so that we update the model after the browser has changed the value.
This obviously runs parsers / validators again.
It also causes some unexpected effects when you bind the same model to range and number for example. When you try to clear the number input, the range will always set the model (and input) back to 50.

As an alternative, I though of creating a validator that adds a misMatch error the the control that is set whenever the browsers updates the value and it is now distinct from the model.

gkalpak commented:
BTW, it doesn't set it to 50, it sets it to `(max - min) / 2`. So, basically, you can't set it to an undefined value. I am still not sure it is a good idea to add something like this is core - too many quirks :worried:

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1436,6 +1481,23 @@ function numberInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+
+  ctrl.$render = function() {
+    if (ctrl.$isEmpty(ctrl.$viewValue)) {
+      ctrl.$viewValue = '50';
```
This shouldn't be hard-coded to 50. It should be `(max - min) / 2`.

gkalpak replied:
BTW, i case it wasn't clear, `50` is the result of `(max - min) / 2` for the default min (`0`) and max (`100`) values.

Narretz replied:
Yep, the PR is currently missing the min max handling.

Narretz replied:
the correct formula is actually `min + ((max - min) / 2)`

gkalpak replied:
Sorry, I meant `(max + min) / 2` (which is equivalent to your formula).
I will correct it to avoid confusion.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,313 @@ describe('input', function() {
+    var scope;
+
+    beforeEach(function() {
+      scope = $rootScope;
```
:smile:

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,313 @@ describe('input', function() {
+        inputElm[0].setAttribute('type', 'text');
+      } catch (e) {}
+
+      helper.changeInputValueTo('100X');
```
Since this is not possible to happen, why do we need to test this?

Narretz replied:
It can happen in IE9. Browsers that don't implement the range should still be able to validate the input correctly  (we mimic the validation rules if no validity is present )

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,313 @@ describe('input', function() {
+
+        // To test how browsers that don't implement range behave, we have to change the type, so
+        // that special layout and validation rules don't apply
+        inputElm[0].setAttribute('type', 'text');
```
I think this might introduce some funny behavior that is different from the actuall behavior (not to mention you said IE8 will throw - not that we care :stuck_out_tongue: ).
I suggest we check if the type is not supported (e.g. `inputElm[0].type === 'text'`) and only then continue with the test.

Narretz replied:
This was copied from another test so we can remove the IE8 line.
Your suggestion is good although i wouldn't put it into the spec itself. Rather make it part of the sniffer?

gkalpak replied:
We already use this approach in ngModel (iirc).
The sniffer typically has some "flags" set, do you suggest we test all available input types on start-up and set appropriate flags on sniffer. I wouldn't be very keen on this. A helper function sounds better. Something like this:

```js
function isOfSupportedType(input) {
  return input.getAttribute('type') === input.type;
}
```

gkalpak replied:
We use it in `baseInputType`. We basically check: `input.type === attrs.type`

gkalpak commented:
I think we need to better handle the situation where `min`/`max` change dynamically. The browser will adjust the element's value to the new `min`/`max` values if necessary and we don't account for that.

Narretz commented:
That's true. I noticed that but haven't included it yet. So what do you in general think about updating the model when the browser changes the values?

gkalpak commented:
Yeah, I think we should read the viewValue every time the `min`, `max` values change (and if necessary run parsers/validators).
(Not sure how `step` affects the element's value.)

Narretz commented:
When a step is set, the browser rounds non-matching values to the nearest step ...

gkalpak commented:
So, how does affect `ngModel`? When the model is set to an "off-step" value, the browser will round the value to the nearest step and the change will propagate to the modelValue?

gkalpak commented:
It is a problematic situation either way. For example if we leave the original (off-step) modelValue, the field will be invalid, but a valid value will be shown to the user. And in order to set the modelValue to the corrected, nearest step value, they will have to move the knob to a different value and back, so we can detect a viewValue change and parse that.

Things like that make me think that supporting range is going to cause more headaches than it will solve problems. Users are probably better of using a non-native slider/range input - e.g. as an independant module or from a UI framework (such as ngMaterial). :disappointed:

Narretz commented:
But for simple cases, binding to range is almost as simple as binding to number. So I'm not sure we should bail out of this simply because there are some edge cases.

petebacondarwin commented:
So from reading the HTML5 specs, I think it is fair to say that range is “special"

In the sense that it can ​*never*​ hold an invalid value

That means that we must always update the ngModel $viewValue to a valid value however it is updated (via user interaction or programmatically).

If we agree that this is the case, and document it clearly, then I think we could go ahead with this directive...

I think that we should also, when reacting to model changes, simply compute the appropriate valid value that would work for the control and update the model explicitly ourselves

So in the case of a model that is out of “step”, if we have this range

```
<input type="range" min="0" max="9" step="2" ng-model="obj.val">
```

and we try to set `obj.val` to `2.6`, then we force the model to `2` as part of the formatting.

roryokane commented on src/ng/directive/input.js:
```diff
@@ -1530,6 +1530,11 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
     // TODO(matsko): implement validateLater to reduce number of validations
     if (minAttrType === 'min') {
+      var elVal = element.val();
+      // IE11 doesn't set the el val correctly if the maxVal is less than the element value
```
This comment doesn’t match the surrounding code. Bad copy-paste from [line 1572](https://github.com/angular/angular.js/blob/0dbe342417f8e0357f72314e2375f62ddaaec6ab/src/ng/directive/input.js#L1572)?

gkalpak commented on karma-shared.conf.js:
```diff
@@ -17,7 +17,7 @@ module.exports = function(config, specificOptions) {
       startConnect: true,
       options: {
-        // We need selenium version +2.46 for Firefox 39 and the last selenium version for OS X is 2.45.
+        // We need selenium version +2.46 for 'Firefox' 39 and the last selenium version for OS X is 2.45.
```
???

Is this comment event relevant anymore.?

Narretz replied:
I don't think so. The change is accidental. Sublime CTRL+D and ' in the wrong file

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1031,6 +1031,112 @@ var inputType = {
+   * Since the element value should always reflect the current model value, a range input
+   * will set the bound ngModel expression to the value that the browser has set for the
+   * input element. For example, in the following input `<input type="range" ng-model="model.value">`,
+   * if I set `model.value = null`, the browser will set the input to `'50'`. Angular will then set
```
_I set_  -->  _the user sets_ (?)

Narretz replied:
*the application sets*. For me, stuff that's set on the scope is only the user's action if it comes from an input.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1031,6 +1031,112 @@ var inputType = {
+   *                  adds no native HTML5 validation. It also means the browser won't adjust the
+   *                  element value in case `min` is greater than the current value.
+   * @param {string=} ngMax Takes an expression. Sets the `max` validation to ensure that the value
+   *                  entered is less than `max`. Does not set the `min` attribute and therefore
```
_min_  -->  _max_

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1031,6 +1031,112 @@ var inputType = {
+
+           Model as range: <input type="range" name="range" ng-model="value"
+                          min="{{min}}"  max="{{max}}">
+          <hr>
```
Wrong indentation here and below.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1031,6 +1031,112 @@ var inputType = {
+         </script>
+         <form name="myForm" ng-controller="ExampleController">
+           Model as range: <input type="range" name="range" ng-model="value"
+                          ng-min="10" ng-max="90">
```
_10_  -->  _min_
_90_  -->  _max_

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1436,6 +1542,124 @@ function numberInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+  var minVal = 0,
+      maxVal = 100,
+      supportsRange = ctrl.$$hasNativeValidators && element[0].type === 'range',
+      validity = element[0].validity;
```
It might be better to use `VALIDITY_STATE_PROPERTY`.

gkalpak replied:
Or not...

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1436,6 +1542,124 @@ function numberInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+    }
+  }
+
+  var minAttrType = isDefined(attr.min) ? 'min' : attr.ngMin ? 'ngMin' : false;
```
Don't `ngMin`/`ngMax` set `attr.min/max` anyway?

Narretz replied:
They do. But iirc at this point (linking) it is not yet set, so I can distinguish it based on attr.min / ngMin

gkalpak replied:
I wouldn't rely on default directive priorities for that. Let's be on the safe side with:

```js
isDefined(attr.ngMin) ? 'ngMin' : isDefined(attr.min) ? 'min' : false;
```

<sub>Note that `attr.ngMin`/`min` could be numbers, and `0` would be falsy, so we need `isDefined` in both cases.</sub>

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1436,6 +1542,124 @@ function numberInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+      } :
+      // ngMin doesn't set the min attr, so the browser doesn't adjust the input value as setting min would
+      function minValidator(modelValue, viewValue) {
+        return ctrl.$isEmpty(viewValue) || isUndefined(minVal) || viewValue >= minVal;
```
Why calling `$isEmpty` with the `viewValue`?

Narretz replied:
Because we are validating the $viewValue. I think dates use the modelValue, but this is something we should change, imo.

gkalpak replied:
I think we almost always call `$isEmpty` with the `modelValue` (regardless of what value we are actually validating). Imo it makes sense to (a) be consistent what we call `$isEmpty` with and (b) always call it with `modelValue` (which can be `$modelValue` or `$$rawModelValue`), in order to account for custom parsers.

For example, an `input[range]`'s viewValue can never be empty, but I might want to have a custom parser that converts `'0'` to `null` and I want this to be considered empty.

--
<sub>It is debatable whether we should be validating the `$viewValue` or the `$modelValue`. I think there are valid usecases for both.</sub>

Narretz replied:
Yes, there are definitely both approaches in the code base. But for example the most recent change that touches this, are the 'ng-empty', and 'ng-not-empty' classes, which are set based on the result of `$ctrl.isEmpty(viewValue)`. https://github.com/angular/angular.js/blob/master/src/ng/directive/ngModel.js#L323
I'd rather make all built-in validators use isEmpty(viewValue), because technically all validators validate what the user has entered, even if we parse it a Date or Number.

gkalpak replied:
I haven't thought that through tbh, but I see merit in both cases. Fwiw, the initial intent seems to be to call it against the `$modelValue`, as indicated by the default implementation (which checks for `undefined`, `null`, `NaN` - values that can hardly appeat as `$viewValue`s).

Imo, the way someone decides to "interpret" the value - i.e. through custom parsers - should be taken into account if possible. But, I understand there are "technical" difficulties with this approach, so checking against the `$viewValue` might be preferrable. 

Most importantly, we should be consistent (if we find ourselves needing to call `$isEmpty` with both values, it is probably an indication that we need to rethink the API - maybe have two separate methods) and communicate it well through docs.

Narretz replied:
FWIW, the required validator calls $isEmpty with the viewValue: https://github.com/angular/angular.js/blob/cd2f6d9d3bcdbcaa3d09dfd9cc75a5058d83b44e/src/ng/directive/validators.js#L70

Though I agree that it's a problem in general that validators sometimes validate the viewValue and sometimes the modelValue.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1436,6 +1542,124 @@ function numberInputType(scope, element, attr, ctrl, $sniffer, $browser) {
+      if (minVal > elVal) {
+        element.val(minVal);
+      }
+      ctrl.$setViewValue(element.val());
```
Can't we use `elVal` here?

Narretz replied:
I guess that works

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,436 @@ describe('input', function() {
+      // allow to set a non-number value and will set the value of the input to 50 even when you
+      // change it directly on the element.
+      // Other browsers fall back to text inputs, where setting a model value of 50 does not make
+      // sense if the input value is a string. These browser will mark the input as invalid instead.
```
_browser_  -->  _browsers_ :innocent:

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,436 @@ describe('input', function() {
+
+        scope.$apply('age = null');
+
+        expect(inputElm.val()).toEqual('50');
```
Nitpick: I would change the element's value first, in order to verify that setting the model to `null` indeed sets the element's value to `'50'` (instead of leaving in unchanged for example).

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,436 @@ describe('input', function() {
+    });
+
+
+    it('should validate with undefined viewValue when $validate() called', function() {
```
_called_  -->  _is called_

gkalpak replied:
Tbh, I am not sure what this test is supposed to test.

Narretz replied:
I think that's copied from input[number] with number substituted for range.

gkalpak commented on test/ng/directive/inputSpec.js:
```diff
@@ -2819,6 +2819,436 @@ describe('input', function() {
+      expect(function() {
+        scope.value = 'one';
+        var inputElm = helper.compileInput('<input type="range" ng-model="value" />');
+      }).toThrowMinErr('ngModel', 'numfmt', "Expected `one` to be a number");
```
Inconsistent use of quotes. (@mgol, I have your back :stuck_out_tongue: )

mgol replied:
Yeah, let's use single anyway, especially that we're not validating it yet.

gkalpak commented:
I just found out that you can make an `input[range]` invalid. Changing the `step` at runtime (while having an off-step value) makes it invalid: it sets the `validity.stepMismatch` flag to true (but keeps the previous value).

I think we need some tests with `step`.

Narretz commented:
I left step out, because I wanted to get it to work with min/max first. I'd like to work on step for number/range after this is merged

gkalpak commented:
> I'd like to work on step for number/range after this is merged

I have mixed feelings about this :smiley: I would rather not release `input[range]` with `step` support.

--
Btw, here is a browser inconsistency (probably a Chrome bug) I found out:

According to (my understanding of) [the spec][1], browsers should automatically adjust the `value` whenevr `step` changes, in order to ensure that there is no step mismatch. When changing the `step` to a string value (e.g. `input.step = '2'`), both Chrome and Firefox follow the spec. When using a number (instead of a string), although both browsers convert `step` to its string representation, only Firefox adjust the `value` and makes the element valid again. Chrome leaves the previous `value` and sets the `validity.stepMismatch` error. E.g.:

```html
<input type="range" step="10" />
```
```js
console.log(input.step)   // --> '10'
console.log(input.value)   //  --> '50'
console.log(input.validity.valid)   // --> true

input.step = '15';
console.log(input.step)   // --> '15'
console.log(input.value)   //  --> '45'
console.log(input.validity.valid)   // --> true

input.step = 20;
console.log(input.step)   // --> '20'
console.log(input.value)   //  --> Firefox: '40'  |  Chrome: '45'
console.log(input.validity.valid)   // --> Firefox: true  |  Chrome: false
```

[1]: https://www.w3.org/TR/html5/forms.html#range-state-(type=range)

Narretz commented:
Okay, step support should be in the same release as the basic support. But it doesn't have to be in the same commit.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1594,15 +1592,16 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
   }
 
-  var minAttrType = isDefined(attr.min) ? 'min' : attr.ngMin ? 'ngMin' : false;
+  var minAttrType = isDefined(attr.ngMin) ? 'ngMin' : isDefined(attr.min) ? 'min' : false;
```
The same change is necessary for `max` below.

gkalpak commented:
LGTM (the travis failures are not related to this PR and have been fixed on master).

Narretz committed:
feat(input): add support for binding to input[type=range]

Thanks to @cironunes for the initial implementation in https://github.com/angular/angular.js/pull/9715

Adds support for binding to input[range] with the following behavior / features:

- Like input[number], it requires the model to be a Number, and will set the model to a Number
- it supports setting the min/max values via the min/max and ngMin/ngMax attributes
- it follows the browser behavior of never allowing an invalid value. That means, when the browser
converts an invalid value (empty: null, undefined, false ..., out of bounds: greater than max, less than min)
to a valid value, the input will in turn set the model to this new valid value via $setViewValue.
-- this means a range input will never be required and never have a non-Number model value, once the
ngModel directive is initialized.
-- this behavior is supported when the model changes and when the min/max attributes change in a way
that prompts the browser to update the input value.
-- ngMin / ngMax do not prompt the browser to update the values, as they don't set the attribute values.
Instead, they will set the min / max errors when appropriate
- browsers that do not support input[range] (IE9) handle the input like a number input (with validation etc.)

Closes #5892
Closes #9715
Close #14870

Narretz commented:
Thanks for the review, @gkalpak ! And thanks again @cironunes for the initial implementation.

mpromonet commented:
Maybe I misunderstand, but even if it changes the behaviour, this PR doesnot fix the issue #6726.   
Using a simple sample from [How to initialize the value of an input[range] using AngularJS when value is over 100](http://stackoverflow.com/q/26453634/3102264) : 
```
    <html ng-app="App">
    <head>
    	<script src="https://code.angularjs.org/snapshot/angular.min.js"></script>
    	<script>
    		angular.module('App', ['App.controllers']);
    		angular.module('App.controllers', []).controller('AppController', function($scope) {
    			$scope.model = {'min':50, 'max':150, 'value':150};
    		});		
    	</script>
    </head>
    <body ng-controller="AppController" >
    	{{model.min}}<input type="range" min="{{model.min}}" max="{{model.max}}"   ng-model="model.value" value="{{model.value}}"  />{{model.max}}<br/>
    	value:{{model.value}}
    </body>
    </html>
```
Before the commit 9130166767c4792c5d32d08a918fc7becf32c9a6 it gives : 

> value = 150 but the slider is at 100.

After it gives :
 
> value = 100 but the slider is at 100.

The 2 views of value are now consistent, but the value and the slider should be at 150 as it is in the range [50,150].

Narretz commented:
@mpromonet Or it's a bug ... (it's a bug). Thanks for testing and reporting.
